engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java

```diff
@@ -28,6 +29,23 @@
  * <p>See also: Note [IR Equality and hashing]
  */
 public interface IR {
+  /**
+   * Compares IR structure, but not metadata neither diagnostics. A special comparator used by
+   * <em>Persistance API</em> to perform some rare consistency checks.
+   */
+  public static final Comparator<IR> STRUCTURE_COMPARATOR =
+      (aIr, bIr) -> {
+        if (aIr == bIr) {
+          return 0;
+        }
+        var aCopy = aIr.duplicate(true, false, false, true);
+        var bCopy = bIr.duplicate(true, false, false, true);
```

Comment:
That looks expensive if we do it for every IR that we read?

Reply:
We are not doing this for every `IR` we read. The code is only performed before exception:
```java
   sb.append("Adding at ").append(at).append(" object:\n  ");
      dumpObject(sb, res);
      sb.append("\nbut there already is:\n  ");
      dumpObject(sb, prev);
      sb.append("\nare they equal: ").append(bothObjectsAreTheSame);
```
is logged or thrown. How frequently that happens? It happens on CI from time to time. I haven't seen it on my computer yet.